v1/remote/write.go

```diff
@@ -34,5 +38,227 @@ type WriteOptions struct {
 // Write pushes the provided img to the specified image reference.
 func Write(ref name.Reference, img v1.Image, auth authn.Authenticator, t http.RoundTripper,
 	wo WriteOptions) error {
-	return fmt.Errorf("NYI: remote.Write(%v)", ref)
+	tr, err := transport.New(ref, auth, t, transport.PushScope)
+	if err != nil {
+		return err
+	}
+	w := writer{
+		ref:     ref,
+		client:  &http.Client{Transport: tr},
+		img:     img,
+		options: wo,
+	}
+
+	bs, err := img.BlobSet()
+	if err != nil {
+		return err
+	}
+
+	// Spin up go routines to publish each of the members of BlobSet(),
+	// and use an error channel to collect there results.
+	errCh := make(chan error)
+	defer close(errCh)
+	for h := range bs {
+		go func(h v1.Hash) {
+			errCh <- w.uploadOne(h)
+		}(h)
+	}
+
+	// Now wait for all of the blob uploads to complete.
+	var errors []error
+	for _ = range bs {
+		if err := <-errCh; err != nil {
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
```

Comment:
This feels a bit weird. Maybe append all the errors together and return a combined one?

https://github.com/hashicorp/go-multierror for inspiration.

Reply:
I only wait for all of them to avoid writes on a closed channel :P